### PR TITLE
Remove mypy exclusions from pre-commit config

### PR DESCRIPTION
**Pull request by tjs-intel**
This exercise highlights a few mypy typing issues in sigopt-server.

 1. Monkeypatching is bad. We patch protobuf's `Message` class with a few auxiliary methods like `copy_protobuf` and `GetFieldOrNone`, but mypy and pylint don't pick up these methods. We should instead use a proxy class or even better just use these as functions instead of methods.
 2. Mypy doesn't like classes that have different types for class variables and instance variables. This is normally not an issue, because we normally use one or the other, but SQLAlchemy's ORM relies on this fact to define models as a collection of `Column` objects. When a class attribute is accessed, it is accessed as a `Column`, ex. `Observation.data.reported_failure`. When the same instance attribute is accessed it returns a native type (ex. `observation.data.reported_failure` is a `bool`). So when mypy encounters a statement like `Observation.data.reported_failure.as_boolean()` it will complain that `bool` has no attribute `as_boolean`. It is a TODO to fix the explicit `.as_*` conversions on protobuf attributes by inferring the types form descriptors, which would fix this specific issue.
 3. `Columns` can define a `proxy` class, which wraps the returned data from a column, but mypy does not know this. We use this ex. for experiment data to provide auxiliary functionality like fetching the sorted parameters. We can easily remedy this issue by moving these extra functionalities elsewhere, or by explicitly defining the type of the column as the proxy type, but I think I tried the latter with no luck.
 4. Defining class attributes as null-able is not good for type checking (and also not a good code practice if they really are required). We see this all over the place, most notably in the handler classes where required objects aren't assigned to the appropriate class properties until after permission checks are performed. This can be remedied by collecting the required objects/data before initializing the class, ex in the handler classes by fetching objects and running permission checks in classmethods, or in services (ex. database service) by creating a separate wrapper around session objects.
 5. We run a bunch of tests by passing invalid types into functions. Pragmatically this is fine, but ideally types are checked statically and therefore we wouldn't need to execute runtime tests of this nature.
 6. Functions like `find` and `dict.get` can return `None` and we often don't check that the result is valid.
 7. Looks like we can't run mypy in pre-commit.ci now that we're checking protobufs

### CODE REVIEW

File: test/integration/v1/constants.py [hunk 1]
```diff
@@ -1,10 +1,154 @@
...
+  metric: NotRequired[AnyMetricMetaType | None]
+  metrics: NotRequired[list[AnyMetricMetaType] | None]
+  observation_budget: NotRequired[int | None]
+  project: NotRequired[str | None]
+  tasks: NotRequired[list[TaskMeta]]
+
+
+class AiExperimentMetaType(BaseExperimentMetaType):
+  metrics: list[AnyMetricMetaType]
+  budget: NotRequired[int | None]
```
Comment (tjs-intel) [inline test/integration/v1/constants.py:148]: This was pretty experimental, but it could be a great way to help validate inputs in the python client.

**Comment by dwanderson-intel** [comment]
I'll take a more thorough look later, but quick high-level comment is that I know we talked about removing protobufs at one point - either the heavy way (migrating many things to columns: more correct, but more tedious espec. in prod) or the light way (interacting with JSONB types more natively, which I think is just a Python dict?). 
  - are either of those feasible? Is the latter "somewhat easy"?
  - would they solve many of the above-related protobuf-mypy issues, or are we likely to run into a whole other slew of them (eg dict access always needs to do a `.get()` with a `None` check to make mypy happy, so we're just trading one headache for another)?

**Comment by tjs-intel** [comment]
> are either of those feasible? Is the latter "somewhat easy"?

I don't really know. I think the latter might involve either replacing the custom protobuf columns with similar custom jsonb wrappers, else we might need to update the code in-line for all the places that we create queries from protobuf access. The former would be a huge undertaking.

**Comment by tjs-intel** [comment]
> would they solve many of the above-related protobuf-mypy issues, or are we likely to run into a whole other slew of them (eg dict access always needs to do a .get() with a None check to make mypy happy, so we're just trading one headache for another)?

Protobuf is actually kind of nice in the typing department. It generates types for mypy to pick up, so we can reliably access required properties of protobuf objects without explicitly checking for `None`. The version of protobuf that we use now seems to validate the data for protobuf classes on instantiation, so we can always be sure that the type information from mypy is correct.

**Review by Chris-Sigopt — APPROVED** [review]
LGTM

**Review by dwanderson-intel — APPROVED** [review]

**Pull request merged**